### PR TITLE
Refactor Song Selection and Song Item

### PR DESCRIPTION
**Pull request by Psycast**
- Changes the SongItem appearance, moves difficulty to the left of the song name.
- Renames paneItems to songItems, uses Vector instead of Array.
- Removes SongItemLocked.as, condensed into SongItem
- Removes 200ms off loading the entire [All] genre compared to Noctowl Editon (1200ms -> 1000ms)
- Removes old removeEventListener unused calls.
- Simplifies keyboard logic to use new isLocked.
- Improves Legacy Filter Logic, only run when filter is enabled instead of always.

### CODE REVIEW

File: src/menu/SongItem.as [hunk 1]
```diff
@@ -1,131 +1,254 @@
...
 
-            //- Diff
-            diffText = new Text(_song["difficulty"], 14);
-            diffText.x = 370;
-            diffText.setAreaParams(_song["difficulty"] >= 100 ? 30 : 27, 27, Text.RIGHT);
-            box.addChild(diffText);
+            switch (songData["access"])
+            {
+                case GlobalVariables.SONG_ACCESS_CREDITS:
+                    return sprintf(_lang.string("song_selection_banned_credits"), {more_needed: NumberUtil.numberFormat(songData.credits - _gvars.activeUser.credits), user_credits: NumberUtil.numberFormat(_gvars.activeUser.credits), song_price: NumberUtil.numberFormat(songData.credits)});
```
Comment (Zageron) [inline src/menu/SongItem.as:205]: Please make this line less huge thank

File: src/menu/SongItem.as [hunk 2]
```diff
@@ -1,131 +1,254 @@
...
+        public function draw():void
         {
-            this.removeEventListener(MouseEvent.ROLL_OVER, boxOver);
-            this.removeEventListener(MouseEvent.ROLL_OUT, boxOut);
-
-            //- Remove is already existed.
-            if (box != null)
+            this.graphics.clear();
+            this.graphics.lineStyle(1, 0xFFFFFF, (highlight ? 0.8 : 0.55));
+            this.graphics.beginGradientFill(GradientType.LINEAR, [0xFFFFFF, 0xFFFFFF], (highlight ? [0.35, 0.1225] : [0.2, 0.04]), [0, 255], Constant.GRADIENT_MATRIX);
```
Comment (Zageron) [inline src/menu/SongItem.as:63]: Lines should not run off the page, please separate highlight tertiary to separate line.

File: src/menu/MenuSongSelection.as [hunk 3]
```diff
@@ -1609,14 +1568,14 @@ package menu
...
-                paneItems[last].active = false;
+            songItems[index].active = true;
+            if (last >= 0 && last < songItems.length)
+                songItems[last].active = false;
 
             // Scroll when doScroll is set.
             if (doScroll && scrollbar.draggerVisibility)
             {
-                var scrollVal:Number = (((paneItems[index].y / pane.content.height) > 0.5) ? ((paneItems[index].y + paneItems[index].height) / pane.content.height) : ((paneItems[index].y) / pane.content.height));
+                var scrollVal:Number = (((songItems[index].y / pane.content.height) > 0.5) ? ((songItems[index].y + songItems[index].height) / pane.content.height) : ((songItems[index].y) / pane.content.height));
```
Comment (Zageron) [inline src/menu/MenuSongSelection.as:1578]: shinky

File: src/menu/MenuSongSelection.as [hunk 4]
```diff
@@ -1064,8 +1022,10 @@ package menu
                     _gvars.loadHighscores(songDetails.level);
                 }
             }
+
+            // Song Details
             else
-            { // Song Details
+            {
                 infoRanks = _gvars.activeUser.getLevelRank(songDetails) || {};
                 var infoDisplay:Array = [[_lang.string("song_selection_song_panel_song"), songDetails['name']], [_lang.string("song_selection_song_panel_author"), songDetails['author']], [_lang.string("song_selection_song_panel_stepfile"), songDetails['stepauthor']], [_lang.string("song_selection_song_panel_length"), songDetails['time']], [_lang.string("song_selection_song_panel_style"), songDetails['style']], [_lang.string("song_selection_song_panel_best"), (infoRanks.score > 0 ? "\n" + NumberUtil.numberFormat(infoRanks.score) + "\n" + infoRanks.results : _lang.string("song_selection_song_panel_unplayed"))]];
```
Comment (Zageron) [inline src/menu/MenuSongSelection.as:1030]: pls

File: src/Constant.as [hunk 5]
```diff
@@ -99,6 +101,17 @@ package
         public static const TEXT_FORMAT_UNICODE_12:TextFormat = new TextFormat(Language.UNI_FONT_NAME, 12, 0xFFFFFF, true);
         public static const JUDGE_WINDOW:Array = [{t: -118, s: 5, f: -3}, {t: -85, s: 25, f: -2}, {t: -51, s: 50, f: -1}, {t: -18, s: 100, f: 0}, {t: 17, s: 50, f: 1}, {t: 50, s: 25, f: 2}, {t: 84, s: 25, f: 3}, {t: 117, s: 0}];
```
Comment (Zageron) [inline src/Constant.as:102]: pls multiline this

File: src/menu/SongItem.as [hunk 6]
```diff
@@ -1,131 +1,254 @@
...
+                _lblMessageText.x = 5;
+                _lblMessageText.y = 20;
+                _lblMessageText.selectable = false;
+                _lblMessageText.embedFonts = true;
+                _lblMessageText.antiAliasType = AntiAliasType.ADVANCED;
+                _lblMessageText.multiline = true;
+                _lblMessageText.width = 395;
+                _lblMessageText.wordWrap = true;
+                _lblMessageText.autoSize = TextFieldAutoSize.LEFT;
+                _lblMessageText.htmlText = "<font face=\"" + Language.UNI_FONT_NAME + "\" color=\"#FFFFFF\" size=\"10\"><b>" + _message.split("\r").join("") + "</b></font>";
```
Comment (Zageron) [inline src/menu/SongItem.as:142]:
```suggestion
                _lblMessageText.htmlText =
                    "<font face=\"" +
                    Language.UNI_FONT_NAME +
                    "\" color=\"#FFFFFF\" size=\"10\"><b>" +
                    _message.split("\r").join("") +
                    "</b></font>";
```

**Comment by Psycast** [comment]
Unable to make suggested changes due to formatting rules.

**Comment by Zageron** [comment]
> 
> 
> Unable to make suggested changes due to formatting rules.

I mean, the tertiary expansion could be done.

**Pull request merged**